proposals/fibers/Explainer.md

```diff
@@ -0,0 +1,874 @@
+
+`handle` blocks may be nested; in which case, the available `event` blocks are the concatenation of all the `event` blocks in all the `handle` blocks that a given switching instruction is textually enclosed within&mdash;up to and including the function body itself.
+
+### Fiber instructions
+
+#### `fiber.new` Create a new fiber
+
+The `fiber.new` instruction creates a new fiber. The instruction has a literal operand which is the index of a function of type `[[ref fiber  r*] t*]->[r*]`, together with corresponding values on the argument stack.
+
+The result of the `fiber.new` instruction is a `fiber` which is the identifier for the newly created fiber. The identity of the fiber is also the first argument to the fiber function itself&mdash;this allows fibers to know their own identity in a straightforward way.
+
+The return values that the fiber function returns represent the return values of the fiber. This is reflected in the type signatures of the `fiber` and the function.
+
+The new fiber is created in a suspended state: it must be the case that the function referenced has one or more `event` blocks within the top-level block of that function. When the fiber is resumed, one of those `event` blocks will be entered as a result of the `fiber.resume`&mdash;depending on the actual event used to resume the fiber.
+
+>If a function is referenced in a `fiber.new` instruction has no `event` blocks at the toplevel, then this function will cause a trap when attempting to resume the new fiber&mdash;because, without any `event` blocks, the fiber will not be able to respond to the resume.
+
+#### `fiber.spawn` Create and enter a new fiber
+
+The `fiber.spawn` instruction creates a new fiber and immediately enters it. It requires a literal function operand of type `[[fiber r*] t*] -> [r*]`, together with values corresponding to `t*` on the value stack.
+
+The `fiber.spawn` instruction is analogous to a `call` instruction, the effect of the instruction is to leave on the value stack the return values of the fiber's function. However, as with other fibers, the new fiber may suspend; in which case there should be appropriate `event` blocks in the instruction stream enclosing the `fiber.spawn` instruction.
+
+Notice that only the newly created fiber has access to the reference that identifies the new fiber. If it is desired that the parent has access to this the code generator would typically insert a `fiber.suspend` instruction at the start of the function and pass out the `fiber` in an appropriate event description.
+
+#### `fiber.suspend` Suspend an active fiber
+
+The `fiber.suspend` instruction takes a fiber as an argument and suspends the fiber. The identified fiber must either be the `active` fiber, or a resume ancestor of the active fiber. 
+
+The _root_ ancestor fiber does not have an explicit identifier; and so it may not be suspended.
```

Comment:
Not being able to suspend the root fiber is a correctness requirement: otherwise, one could end up in a situation where no forward execution is available.
However, this does not clash with JSPI. In that world, the export call creates a fiber for the export to run in. This is a 'local' root but not the global root (which is typically connected to the browser's event loop)

Reply:
The problem is that that "local" root does not have an identity because from this proposal's POV it is still the root, so it cannot be suspended by this proposal. So the call to promise-returning import would still be necessary. This could be fixed by giving root fibers identity so they can be suspended. Yes, execution would return to the host at that point, but that doesn't seem like a problem. In particular, suspending all Wasm forward progress and returning control to the host is exactly what JSPI does.

Reply:
Actually, if you are using JSPI, then your local root will have identity and you will be able to suspend it. And, as you pointed out, that will cause the entire app to suspend and 'return' to the browser.